streams: Fixes issue in stream creation.

Ujjawal3 opened the pull request:
This PR fixes issue in stream creation.
<!-- Describe your pull request here.-->

[CZO discussion](https://chat.zulip.org/#narrow/stream/9-issues/topic/Issue.20in.20stream.20creation)

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**


Stream Creation form when user cannot subscribe other users.
![Screenshot from 2023-05-11 13-23-55](https://github.com/zulip/zulip/assets/99073049/d290be46-455c-4176-aa69-87ec2916cc2d)
![Screenshot from 2023-05-11 13-23-52](https://github.com/zulip/zulip/assets/99073049/0585cdaa-e60a-4314-bdf7-bf641123c601)


Stream edit when user cannot subscribe other users.
![Screenshot from 2023-05-11 13-24-11](https://github.com/zulip/zulip/assets/99073049/f954d24c-af36-4ea7-9ecb-ea1ee30b8e84)


<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

tungmeoo reviewed (commented):
yêu cầu kéo

Ujjawal3 commented on zerver/tests/test_subs.py:
```diff
@@ -4722,27 +4722,49 @@ def test_bulk_subscribe_MIT(self) -> None:
-        with self.tornado_redirected_to_list(events, expected_num_events=0):
-            with self.assert_database_query_count(5):
+        with self.tornado_redirected_to_list(events, expected_num_events=41):
+            with self.assert_database_query_count(52):
```
Earlier user was not subscribed because of the wrong condition `len(principals)>0` in view function used for adding subscriptions. Now after changing the condition user is successfully subscribed to streams because of which query count increases by this much.

timabbott replied:
41 events and 52 databases calls sounds like a performance problem.

But the reason for the Zephyr detail is that the security model for Zephyr (the protocol we're mirroring with) is that stream names have a security-through-obscurity model; anyone can subscribe if they know the name.

So we definitely intend not to send any events to other users when adjusting subscriptions in such a stream.

timabbott replied:
Reading this, I cannot figure out what change in your code caused this to generate 41 events; I would have expected  a single event here.

Ujjawal3 replied:
The increase in query counts and events is because of calling this function while adding subscriptions.
```
def send_stream_creation_events_for_private_streams(
    realm: Realm,
    stream_dict: Dict[int, Stream],
    altered_user_dict: Dict[int, Set[int]],
) -> None:
    for stream_id, stream_users_ids in altered_user_dict.items():
        stream = stream_dict[stream_id]

        if not stream.is_public():
            # Users newly added to invite-only streams
            # need a `create` notification.  The former, because
            # they need the stream to exist before
            # they get the "subscribe" notification, and the latter so
            # they can manage the new stream.
            # Realm admins already have all created private streams.
            realm_admin_ids = {user.id for user in realm.get_admin_users_and_bots()}
            notify_user_ids = list(stream_users_ids - realm_admin_ids)

            if notify_user_ids:
                send_stream_creation_event(stream, notify_user_ids)
```
For this test if block is executed as `stream.is_public()` returns `False` because of stream is in `zephyr_realm`. I think 
we should change the if condition to `stream.invite_only` and also move `realm_admin_ids` outside the loop.

timabbott replied:
It appears these are `stream_created` events that get sent when the user themself joins the stream; I think that probably is correct that they get those.

So the fix here is twofold:
* We should explain that these events are from `send_stream_creation_events_for_private_streams` in the comment.
* We should assert that the events have the right `type`

timabbott replied:
OK rewrote the test to explain the behavior clearly.

Ujjawal3 commented:
@sahil839 please review this PR.

sahil839 commented on web/src/stream_ui_updates.js:
```diff
@@ -228,22 +229,37 @@ export function update_add_subscriptions_elements(sub) {
+    // We should disable the widgets when stream is private or
+    // realm level setting does not allow user to add subscribers.
+    const allow_user_to_add_subs =
+        sub.can_add_subscribers && settings_data.user_can_subscribe_other_users();
```
I think this check of `user_can_subscribe_other_users()` should be added to `stream_data.can_subscribe_others` instead of adding it here.

Ujjawal3 replied:
Do we want a common tooltip message for all cases? At present there are two different tooltip messages depending on whether adding subscribers was disabled because stream was private or because of realm level setting.

alya replied:
What are the two messages?

Ujjawal3 replied:
As per present implementation if stream is private and user is not subscribed to stream we show `Only stream members can add users to a private stream` and if user is not allowed to add subscribers because of realm level settings we show `You are not allowed to add subscribers to stream.`

alya replied:
Hmm, I think it's helpful to keep those two separate -- the first one provides good info for org admins.

There should be a period after the first one, and let's change the second one to, "You are not allowed to add subscribers to this stream." for better grammar.

sahil839 commented on web/src/stream_ui_updates.js:
```diff
@@ -228,22 +229,37 @@ export function update_add_subscriptions_elements(sub) {
         $input_element.popover("destroy");
     } else {
-        $input_element.prop("disabled", true);
+        $input_element.prop("contenteditable", false);
```
I think this can be a prep commit as this does not seem related to this. User should not have been able to have the cursor inside the input when button is disabled.

sahil839 reviewed (commented):
Posted some comments. I think would be better to properly describe bug in the commit message for first commit.

I am not sure we discussed about hiding or disabling the add subscribers widget in stream creation form. I like the idea of hiding it, but am not sure what is the best solution. @alya FYI.
And in case we are hiding the widget, the `Filter subscribers` input is not aligned properly wit the heading when the add subscribers widget is hidden.

alya commented:
> I am not sure we discussed about hiding or disabling the add subscribers widget in stream creation form. I like the idea of hiding it, but am not sure what is the best solution. @alya FYI.

Hm, I think we need some way to communicate to the user that they don't have permissions to add subscribers. It's probably fine to disable the user input field, "Add" button, and the "Add all users" button, with the same tooltip as on the the "Edit stream" panel. That might be simplest, and then we won't have to adjust the filter location?

sahil839 commented:
@Ujjawal3 can you update the PR to fix the tests and post a comment when it is ready for review? Would be good to fix this issue soon.

Ujjawal3 commented:
> @Ujjawal3 can you update the PR to fix the tests and post a comment mentioning the updates when it is ready for review? Would be good to fix this issue soon.

I was busy with my college exams since last week so could not do much contributions. But tomorrow they will finish so will work after that.

Ujjawal3 commented:
@sahil839 I have updated the PR and addressed all your comments. Also now add subscriber section is disabled in stream creation form. Please have a look.

sahil839 commented on web/src/stream_create.js:
```diff
@@ -360,7 +360,34 @@ export function show_new_stream_modal() {
+    if (settings_data.user_can_subscribe_other_users()) {
+        $add_subscribers_container.removeClass("add_subscribers_disabled_tooltip");
+
+        $input_element.prop("contenteditable", true);
```
Some part of this is duplicated from the code in `stream_ui_updates.update_add_subscriptions_elements`. but I don't think it would help much to reuse it. But would be worth looking at and we can do it if it does not affect readability much.

Ujjawal3 replied:
I cannot use `stream_ui_updates.update_add_subscription_elements` because this function makes changes in the UI of existing streams not in the UI creating new streams. But I liked the second approach and updated the PR.

sahil839 replied:
Ok, I was thinking of passing container element to it. But now I think a better approach would be to create a new function like `enable_or_disable_add_subscribers_elements` with `container_elem` (and probably `is_stream_creation` too) as paramters and then call this both from `stream_ui_updates.update_add_subscription_elements` and this.

And this function would only handle enabling and disabling of elements and probably tooltips as well if it does not affect readability.

This change might require some small prep commits to keep it easy to review.

sahil839 reviewed (commented):
Thanks for updating the PR. Posted few comments.

Ujjawal3 commented on zerver/tests/test_subs.py:
```diff
@@ -4732,26 +4732,48 @@ def test_bulk_subscribe_MIT(self) -> None:
+        self.assert_json_error(
+            result,
+            "You can only invite other Zephyr mirroring users to private streams.",
+            status_code=400,
```
I think this should be in same commit because removing this causes the commit making changes in backend  to fail some tests as there would not be 100% test coverage. This test was added to fix the failing backend tests.

sahil839 replied:
You can probably add this test in a commit before this one mentioning that the case of subsribing others to public streams was not tested. Since in the `test_bulk_subscribe_MIT` test, the user subscribes themselves and not anyone else.

sahil839 reviewed (commented):
Posted a couple of comments.

Also, regarding the commit messages, we generally use "refactor" only when just the structure of code changes and behavior remains same. But the first commit fixes a bug and changes the behavior of the API, so you might want to update the commit message as well.

Ujjawal3 commented:
@sahil839 I have done the changes as per your comments. Please have a look.

sahil839 reviewed (commented):
Posted a couple of comments. @alya do we want to include this in this release? We can probably skip the CSS changes mentioned in [this comment](https://github.com/zulip/zulip/pull/24900#discussion_r1187293668) for now and do it as a follow-up, but we might still want to fix the cursor for disabled button and input.

alya commented:
In addition to the cursor fix, the tooltip placement looks odd. I don't have strong feeling about the tooltip on the "Add" button, but the tooltip on the "Add all users" button should be attached to that button, not to the text field above it.

alya commented:
I think its not *critical* (this combination of permissions is probably rare, though I haven't checked), but would be nice not to have this bug in the release.

Ujjawal3 commented on web/src/stream_ui_updates.js:
```diff
@@ -286,3 +280,18 @@ export function update_setting_element(sub, setting_name) {
+    $add_subscribers_button.prop("disabled", !enable_elem);
+
+    if (enable_elem) {
+        $add_subscribers_button.css("pointer-events", "");
```
@sahil839 I have updated the PR to show the "not-allowed" cursor for disabled field and disabled input and changed the position of tooltip for "Add all users" button.
 Regarding CSS properties changed through Js  `initialize_disable_btn_hint_popover` is called at around 5 places in code. So for doing changes in this function we need to make changes in all those places. So, should I do the changes inside the function ?

sahil839 replied:
We show "not-allowed" cursor in the space between the input and "Add" button as well. Not sure how important it is as there is very less space. 

I think we can do the "pointer-events" change as follow-up, since it was present before this PR as well.

sahil839 commented:
Code changes look good to me. @alya do you have any more feedback here?

timabbott commented:
I resolved the feedback; merging this as it fixes the bug and I think it is pretty low risk that this is worse; probably still worth @alya doing a final check sometime before the 7.0 final release in case there's further tweaks to do.

Thanks @Ujjawal3!

Ujjawal3 commented:
@timabbott PR was not merged because of CI failures due to linting errors. I have updated the PR. Please merge it now.

alya commented:
Thanks! Could you also note in the PR description / commit message that this resolves #25546 (according to @sahil839 's [investigation](https://chat.zulip.org/#narrow/stream/9-issues/topic/bug.20in.20create.20stream.20UI/near/1568727)).

timabbott commented:
Adjusted the commit message and merged, thanks @Ujjawal3!

alya commented:
@Ujjawal3 Why is the styling on the tooltip in the stream edit UI totally different from other tooltips? Can we change it to match?

alya commented:
Looks good to me otherwise.
